payments_payu/provider.py

```diff
@@ -593,6 +603,90 @@ def process_data(self, payment, request, *args, **kwargs):
+        if ext_refund_id is not None:
+            request_data.setdefault("refund", {}).setdefault(
+                "extRefundId", ext_refund_id
+            )
+
+        response = self.post_request(request_url, data=json.dumps(request_data))
+
+        payment_extra_data = json.loads(payment.extra_data or "{}")
+        payment_extra_data_refund_responses = payment_extra_data.setdefault(
+            "refund_responses", []
+        )
+        payment_extra_data_refund_responses.append(response)
+        payment.extra_data = json.dumps(payment_extra_data, indent=2)
+        payment.save()
+
+        try:
+            refund = response["refund"]
+            refund_id = refund["refundId"]
+        except Exception:
+            refund_id = None
+
+        try:
+            response_status = dict(response["status"])
+            response_status_code = response_status["statusCode"]
+        except Exception:
+            raise ValueError(
+                f"invalid response to refund {refund_id or '???'} of payment {payment.id}: {response}"
+            )
+        if response_status_code != "SUCCESS":
+            raise ValueError(
+                f"refund {refund_id or '???'} of payment {payment.id} failed: "
+                f"code={response_status.get('code', '???')}, "
+                f"statusCode={response_status_code}, "
+                f"codeLiteral={response_status.get('codeLiteral', '???')}, "
+                f"statusDesc={response_status.get('statusDesc', '???')}"
+            )
+        if refund_id is None:
+            raise ValueError(
+                f"invalid response to refund of payment {payment.id}: {response}"
+            )
+
+        try:
+            refund_order_id = response["orderId"]
+            refund_status = refund["status"]
+            refund_currency = refund["currencyCode"]
+            refund_amount = dequantize_price(refund["amount"], refund_currency)
+        except Exception:
+            raise ValueError(
+                f"invalid response to refund {refund_id} of payment {payment.id}: {response}"
+            )
+        if refund_order_id != payment.transaction_id:
+            raise NotImplementedError(
+                f"response of refund {refund_id} of payment {payment.id} containing a different order_id "
+                f"not supported yet: {refund_order_id}"
+            )
+        if refund_status == "CANCELED":
+            raise ValueError(f"refund {refund_id} of payment {payment.id} canceled")
+        elif refund_status not in {"PENDING", "FINALIZED"}:
+            raise ValueError(
+                f"invalid status of refund {refund_id} of payment {payment.id}"
+            )
+        if refund_currency != payment.currency:
+            raise NotImplementedError(
+                f"refund {refund_id} of payment {payment.id} in different currency not supported yet: "
+                f"{refund_currency}"
+            )
+        return refund_amount
```

Comment:
I don't like that we return the amount before the refund actually happens but this is how PayPal refunds work too so I decided to be consistent...